KC-1029 updating MRC demo with observer promotion (#857)
confluentinc/examples: multiregion/docs/multiregion.rst

topic-partition, and consumers are allowed to read from them even though the
*Observer* isn't the leader–this is known as “Follower Fetching”. However, the
data is copied asynchronously from the leader such that a producer doesn't wait

> on observers to get back an acknowledgement. In non degraded states, observers don't

Reviewer: ```suggestion
on observers to get back an acknowledgement. In "non-degraded" steady state, observers don't
```

|Follower_Fetching|

> ``Observer Promotion`` is the process whereby an observer is promoted into the

Reviewer: @thomaskwscott in light of this sentence & behavior change, does the sentence above need to be reworked?  Specifically referring to: https://github.com/confluentinc/examples/pull/857/files#diff-83774907eea7bccda4f64eb4cb083aa94364c5e5ea244a1a189a744bb2705379R44-R47. It might also need to be partially moved down to this paragraph as well

Author: good point, I've changed the observer description above.

ISR in certain degraded scenarios. This behaviour is controlled by the
``observerPromotionPolicy`` field in a topic's replica placement policy. It can
have values:

> - under-min-isr: observers will be promoted if the isr size drops below the topic's min.insync.replicas configuration.

Reviewer: ```suggestion
- ``under-min-isr``: observers will be promoted if the ISR size drops below the topic's ``min.insync.replicas`` configuration.
```

> - under-replicated: observers will be promoted if the isr size drops below the configured count of replicas in the topic's replica placement policy.

Reviewer: ```suggestion
- ``under-replicated``: observers will be promoted if the ISR size drops below the ``replicas-->count`` value in the topic's replica placement policy.
```

> - leader-is-observer: observers will only be promoted if the leader is an observer.

Reviewer: ```suggestion
- ``leader-is-observer``: observers will only be promoted if the leader is an observer.
```

Reviewer: "If the leader..." -- the current leader that just failed?  I'm a bit confused by this one.

Author: if the current leader is an observer, this can only happen after unclean leader election I believe and is the same as the previous behaviour. I've updated the wording to reflect this.

     - Observers (async replicas)
     - ISR list
     - Use default placement contraints

>      - Promotion policy

Reviewer: ```suggestion
     - Observer promotion policy
```

Author: fixed

   * - single-region
     - 1x west
     - 1x west
     - n/a
     - {1,2}
     - no
     - none

   * - multi-region-sync
     - 1x west
     - 1x west, 2x east
     - n/a
     - {1,2,3,4}
     - no
     - none

   * - multi-region-async
     - 1x west
     - 1x west
     - 2x east
     - {1,2}
     - no
     - none

   * - multi-region-default
     - 1x west
     - 1x west
     - 2x east
     - {1,2}
     - yes
     - none

>    * - under-min-isr-promotion

Reviewer: For consistency in naming, perhaps these new topics should be called:

- multi-region-async-op-under-min-isr
- multi-region-async-op-under-replication (and this one maybe should be `replica-count` instead of `replication`, for consistency with the policy parameter name?)
- multi-region-async-op-leader-is-observer

Reviewer: cc: @thomaskwscott please see above

Author: names changed, I used multi-region-async-op-under-replicated to match the policy name

      multi-region-sync: 4
      multi-region-async: 4
      multi-region-default: 4
      under-min-isr-promotion: 4
      under-replicated-promotion: 4
      leader-is-observer-promotion: 4

Failover and Failback
---------------------

> Degrade Region

Reviewer: We should consider promoting (pun intended) `Degrade Region` into its own section, on the same level as `Failover and Failback`.  

1. Degrade is not a true failover
2. Degrade has merits in its own right

Author: good point, I've moved it up

~~~~~~~~~~~~~~

> In this section, you will simulate a broker failure in the ``west`` region.

Reviewer: ```suggestion
In this section, you will simulate a single broker failure in the ``west`` region.
```

Author: updated

> #. Run the following command to stop the Docker containers corresponding to the ``west`` region:

Reviewer: This statement needs tweaking, not all west region was taken down

Author: good point, I changed to to say that we are killing one broker container in west.

      	Topic: under-replicated-promotion	Partition: 0	Leader: 2	Replicas: 2,1,3,4	Isr: 2,4	Offline: 1	Observers: 3,4

      ==> Describe topic leader-is-observer-promotion

      Topic: leader-is-observer-promotion	PartitionCount: 1	ReplicationFactor: 4	Configs: min.insync.replicas=1,confluent.placement.constraints={"observerPromotionPolicy":"leader-is-observer","version":2,"replicas":[{"count":2,"constraints":{"rack":"west"}}],"observers":[{"count":2,"constraints":{"rack":"east"}}]}
      	Topic: leader-is-observer-promotion	Partition: 0	Leader: 2	Replicas: 2,1,3,4	Isr: 2	Offline: 1	Observers: 3,4

#. Observe the following:

   - In all topics except ``under-min-isr-promotion``, ``multi-region-sync`` and ``under-replicated-promotion``
     there is only 1 replica in the ISR. This is because replica placement dictated all replicas were in the ``west``
     region which has only 1 remaining live broker.

>    - In the second scenario, the ``multi-region-sync`` topic maintained an ISR of 3 brokers. This is because it's

Reviewer: ```suggestion
   - In the second scenario, the ``multi-region-sync`` topic maintained an ISR of 3 brokers. This is because its
```

     placement policy always allows for brokers from east to join the ISR.

>    - The ``under-min-isr-promotion`` and ``under-replicated-promotion`` topics have placement policies that allow

Reviewer: Should this say _automatically_, to highlight that this happens without manual intervention?

Author: fixed

     observers to be promoted into the ISR. In the case of ``under-min-isr-promotion`` the number of non-observer
     replicas (1) is less than the ``min.insync.replicas`` value (2). Observers are promoted to the ISR to meet the
     ``min.insync.replicas`` requirement. In the case of ``under-replicated-promotion`` the number of online replicas
     (1) is less than the intended number of non observer replicas from the replica placement (2). An observer is

>      promoted to fulfil this requirement.

Reviewer: ```suggestion
     promoted to fulfill this requirement.
```

Author: fixed

      multi-region-sync: 4
      multi-region-async: 4
      multi-region-default: 4
      multi-region-async-op-under-min-isr: 4
      multi-region-async-op-under-replicated: 4
      multi-region-async-op-leader-is-observer: 4

      ==> Monitor ObserversInIsrCount

      single-region: 0
      multi-region-sync: 0
      multi-region-async: 0
      multi-region-default: 0
      multi-region-async-op-under-min-isr: 0

>       multi-region-async-op-under-replicated: 0

Reviewer 2: QQ: Should this have a non zero value if we've promoted observers in this example?

Author: yes, but these metrics are shown before region failure and hence no observers have been promoted.

Reviewer: @skaundinya15 I'm reworking the docs with more JMX meric output, so this will become clearer based on when the JMX metrics are captured.  Please stay tuned...PR in progress ;)

Reviewer 2: Makes sense, thanks @thomaskwscott. Looking forward to that @ybyzek 💯

      multi-region-async-op-leader-is-observer: 0

Degraded Region
---------------

In this section, you will simulate a single broker failure in the ``west`` region.

> #. Run the following command to stop on of the broker Docker containers corresponding in the ``west`` region:

Reviewer: ```suggestion
#. Run the following command to stop one of the broker Docker containers in the ``west`` region:
```

     elected a new leader in ``east`` (for example, replica 3 in the previous
     output). Clients can failover to those replicas in the ``east`` region.

   - The ``multi-region-async``, ``multi-region-default`` and
     ``leader-is-observer-promotion`` topics have no leader, because they had
     only two replicas in the ISR, both of which were in the ``west`` region and
     are now down. The observers in the ``east`` region are not eligible to become
     leaders automatically because they were not in the ISR.

   - The ``under-min-isr-promotion`` and ``under-replicated-promotion`` topics have

>      have promoted observers into the ISR and an observer has become the leader.

Reviewer: ```suggestion
    promoted observers into the ISR and an observer has become the leader.
```

>      This is because their observerPromotionPolicy allows this.

Reviewer: ```suggestion
     This is because their replica placement policy has set ``observerPromotionPolicy`` to allow this.
```
